src/main/java/com/lightbend/akka/jpa/sample/UserRepository.java

```diff
@@ -0,0 +1,140 @@
+      Root<User> root = criteria.from(User.class);
+      criteria.select(root);
+
+      List<User> result = entityManager.createQuery(criteria).getResultList();
+      return result;
+    });
+  }
+
+  public CompletionStage<List<User>> findByName(String name) {
+    return withEntityManager(entityManager -> {
+      CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+      CriteriaQuery<User> criteria = builder.createQuery(User.class);
+      Root<User> root = criteria.from(User.class);
+      criteria.select(root).where(builder.equal(root.get("name"), name));
+
+      List<User> result = entityManager.createQuery(criteria).getResultList();
+      return result;
+    });
+  }
+
+  public CompletionStage<Done> save(User user) {
+    return withEntityManager(entityManager -> {
+      if (user.isNew()) {
+        // new user
+        entityManager.persist(user);
+      } else {
+        // update
+        entityManager.merge(user);
+      }
+      return Done.getInstance();
+    });
+  }
+
+  public CompletionStage<Done> delete(long id) {
+    return withEntityManager(entityManager -> {
+      Optional<User> maybeUser = blockingFindById(entityManager, id);
+      if (maybeUser.isPresent())
+        entityManager.remove(maybeUser.get());
+      return Done.getInstance();
+    });
+  }
+
+  /**
+   * All data access should be performed via this method.
+   *
+   * It handles the lifecycle of the EntityManager, i.e. closing it after
+   * the data access code returns or throws.
+   *
+   * A transaction is started before the data access code is executed and
+   * it is committed afterwards  or rolled back if exception was thrown.
+   *
+   * The blocking data access code is executed in a dedicated Akka dispatcher
+   * (thread pool) to avoid starvation of other non-blocking parts of the system.
+   *
+   * The `CompletionStage` is returned immediately and completed later when with the
+   * value that the data access code returns. If exception is thrown by the entity manager,
+   * the data access code, or transaction the `CompletionStage` is completed exceptionally
+   * with that exception.
+   */
+  private <T> CompletionStage<T> withEntityManager(Function<EntityManager, T> dataAccess) {
+    CompletableFuture<T> result = new CompletableFuture<>();
+    blockingDispatcher.execute(() -> {
+      try {
+        final EntityManager entityManager = entityManagerFactory.createEntityManager();
+        try {
+          entityManager.getTransaction().begin();
+          T r = dataAccess.apply(entityManager);
+          entityManager.getTransaction().commit();
+          entityManager.close();
+          result.complete(r);
+        } catch (Exception e) {
+          try {
+            EntityTransaction txn = entityManager.getTransaction();
+            if (txn.isActive())
+              txn.rollback();
+            entityManager.close();
+          } catch (Exception ignore) {
+            // the original exception is more interesting
```

Comment:
It's good practice to add the ignored exception to the suppressed list of the one that is propagated: https://docs.oracle.com/javase/8/docs/api/java/lang/Throwable.html#addSuppressed-java.lang.Throwable- (also handled automatically by try-with-resources)